src/beamlime/applications/_nexus_helpers.py

```diff
@@ -91,29 +91,82 @@ class RunStartInfo:
     TODO: Remove this field when it is no longer used."""
 
 
+class InvalidNexusStructureError(ValueError):
+    """Raised when the nexus structure is invalid."""
+
+
+def _validate_module_configs(
+    key_value_pairs: tuple[tuple[StreamModuleKey, StreamModuleValue], ...],
+) -> None:
+    """Validate the module configuration in the nexus structure."""
+    invalid_module_paths = tuple(
+        "/".join(part for part in value.path if part is not None)
+        for key, value in key_value_pairs
+        if not isinstance(key.topic, str) or not isinstance(key.source, str)
+    )
+    if len(invalid_module_paths) != 0:
+        raise InvalidNexusStructureError(
+            "Invalid module place holder(s) found in the nexus structure: ",
+            "\n".join(invalid_module_paths),
+        )
+
+
+def _validate_module_keys(
+    key_value_pairs: tuple[tuple[StreamModuleKey, StreamModuleValue], ...],
+) -> None:
+    """Validate the module keys."""
+    key_set = {key for key, _ in key_value_pairs}
+    key_list = [key for key, _ in key_value_pairs]
+    for unique_key in key_set:
+        key_list.remove(unique_key)
+    # ``key_list`` should be empty if all keys are unique.
+    if len(key_list) != 0:
+        raise InvalidNexusStructureError(
+            "Duplicate module place holder(s) found in the nexus structure: ", key_list
+        )
```

Comment:
```suggestion
    from collections import Counter
    key_list = [key for key, _ in key_value_pairs]
    key_counts = Counter(key_list)
    if max(key_counts.values()) > 1:
        raise InvalidNexusStructureError(
            "Duplicate module place holder(s) found in the nexus structure: ", key_list
        )
```
Just a styling choice, feel free to ignore :)

Reply:
Oh I just didn't know of this. I'll slightly update it with filtering the counts. 
The idea was that ``key_list`` only contains the duplicating keys for the error message.
Thanks.